### ISSUE TITLE
Add support for functions

### ISSUE THREAD

**Issue by dpsanders** [dpsanders/IntervalConstraintProgramming.jl#21]
Functions from n variables to 1:

``` julia
@function f(x, y) = x^2 + y^2

@constraint 1 <= f(x, y) <= 3

@function f(x) = 4x*(1-x)
@constraint 0.1 <= f(f(x)) <= 0.3
```

**Comment by dpsanders** [comment]
A function like `f(x,y) = x^2 + y^2` from `Rˆn` to `R` needs a forward pass and a reverse pass.
The reverse pass takes in the constraint on the value of `f(x,y)`, together with current values of `x` and `y` (?)  and returns the constrained values of `x` and `y`:

`(x, y) = _reverse_f_(constraint, x, y)`

**Comment by dpsanders** [comment]
This is basically what is already implemented in `contractor.jl`; it needs to be factored out.

**Comment by dpsanders** [comment]
It's just effectively adding a new reverse mode function.

**Comment by dpsanders** [comment]
Equiv to

```
z1 = f(x)
z2 = f(z1)
z2 = z2 ∩ constraint
z1 = _reverse_f_(z2, z1)
x = _reverse_f_(z1, x)
```

**Comment by dpsanders** [comment]
The reverse mode function needs to take as arguments _all_ of the intermediate variables `z1`, `z2` etc., since it needs their current values. So `_reverse_f_` becomes `_reverse_f_(z1, z2, constraint)`

**Comment by dpsanders** [comment]
1D functions implemented in #38.
